### ISSUE TITLE
Always shows provider selection page after upgrading IS4 from 3.1 to 4.0.1.

### ISSUE THREAD

**Issue by cdibbs** [IdentityServer/IdentityServer4#4596]
IdentityServer4 always shows its own provider selection page for existing clients now. When investigating this, I noticed it would re-enter this method:

```csharp
        [HttpGet]
        public async Task<IActionResult> Login(string returnUrl)
        {
            // build a model so we know what to show on the login page
            var vm = await BuildLoginViewModelAsync(returnUrl);

            if (vm.IsExternalLoginOnly)
            {
```

The first time, `vm.IsExternalLoginOnly` would be true, indicating it should skip showing the LoginView. It would visit ExternalController.Challenge, then return to `Login` but this time with `vm.IsExternalLoginOnly == false`.

In the database, I double-checked `[ClientIdPRestrictions]` and all of the clients have only one provider. If I oblige and pick a provider, it throws the following error upon return:

```
2020-06-29 16:28:26.312 -05:00 [DBG] AuthenticationScheme: idsrv.external was not authenticated.
2020-06-29 16:28:26.340 -05:00 [INF] Executed action IdentityServerHost.Quickstart.UI.ExternalController.Callback (AppDev.IdentityServer4) in 1188.1077ms
2020-06-29 16:28:26.341 -05:00 [ERR] An unhandled exception has occurred while executing the request.
System.Exception: External authentication error
   at IdentityServerHost.Quickstart.UI.ExternalController.Callback() in C:\d\git\appdev.identityserver4\AppDev.IdentityServer4\Quickstart\Account\ExternalController.cs:line 90
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.TaskOfIActionResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask`1 actionResultValueTask)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|24_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeFilterPipelineAsync()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)
   at Microsoft.AspNetCore.Builder.RouterMiddleware.Invoke(HttpContext httpContext)
   at IdentityServer4.Hosting.IdentityServerMiddleware.Invoke(HttpContext context, IEndpointRouter router, IUserSession session, IEventService events)
   at IdentityServer4.Hosting.MutualTlsEndpointMiddleware.Invoke(HttpContext context, IAuthenticationSchemeProvider schemes)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at IdentityServer4.Hosting.BaseUrlMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
2020-06-29 16:28:26.415 -05:00 [INF] Request finished in 1264.4551ms 500 text/html; charset=utf-8
```

That seems to be because Callback gets called twice, the first time successfully, then unsuccessfully.

What could I be missing? Thanks in advance.

**Comment by leastprivilege** [comment]
I would put a breakpoint into the ExternalController.Callback method to see what's going on there.

**Comment by cdibbs** [comment]
@leastprivilege Thanks for a quick reply. After returning from logging in, result.Succeeded is true, and I see a valid-looking (and valid/working previously in 3.1) Principal containing the desired user etc. `Callback` makes it all the way down to `Redirect(returnUrl)`, but that's set to "/External/Callback", so it ends up calling itself again. This time, after awaiting AuthenticateAsync, the result is false and full of empty fields.

**Comment by osagga** [comment]
I just stumbled upon this issue, the problem for me was this bug in the quickstart code (all of the examples 1-6) here:
https://github.com/IdentityServer/IdentityServer4/blob/4624ed0aed7f3f3ce9fb2f15147fdf95338e8802/samples/Quickstarts/6_AspNetIdentity/src/IdentityServerAspNetIdentity/Quickstart/Account/AccountController.cs#L61-L62

The `RedirectToAction` assumes that the first route value is called `provider`, when in fact its called `scheme` (this rename change was introduced recently in 4624ed0aed7f3f3ce9fb2f15147fdf95338e8802 and 6924ff0454f02cb3265c2eb251a1f3d387251e1e)

This name mismatch causes the `scheme` in the `Challenge` call to be `null` here:
https://github.com/IdentityServer/IdentityServer4/blob/6924ff0454f02cb3265c2eb251a1f3d387251e1e/samples/Quickstarts/3_AspNetCoreAndApis/src/IdentityServer/Quickstart/Account/ExternalController.cs#L51

I'm not sure why the `scheme` is not checked for `null` in the action, but basically since its null, the challenge call below will just cause a redirect to the login action (now with a `returnUrl` set to `External/Challenge`) which causes a loop of redirects.

https://github.com/IdentityServer/IdentityServer4/blob/6924ff0454f02cb3265c2eb251a1f3d387251e1e/samples/Quickstarts/3_AspNetCoreAndApis/src/IdentityServer/Quickstart/Account/ExternalController.cs#L73

For now I just fixed it by renamming `provider` to `scheme` [here](https://github.com/IdentityServer/IdentityServer4/blob/4624ed0aed7f3f3ce9fb2f15147fdf95338e8802/samples/Quickstarts/6_AspNetIdentity/src/IdentityServerAspNetIdentity/Quickstart/Account/AccountController.cs#L62) (since it seems that's the new intended name for that value), but maybe a null check should be added to make sure that the `scheme` is always set? or maybe just drop the name since its an anonymous type? I can open a PR with this fix, just let me know what's the expected behavior here.
